x-pack/libbeat/management/blacklist.go

```diff
@@ -0,0 +1,188 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/libbeat/management/api"
+)
+
+// ConfigBlacklist takes a ConfigBlocks object and filter it based on the given
+// blacklist settings
+type ConfigBlacklist struct {
+	patterns map[string]*regexp.Regexp
+}
+
+// ConfigBlacklistSettings holds a list of fields and regular expressions to blacklist
+type ConfigBlacklistSettings map[string]string
+
+// Unpack unpacks nested fields set with dot notation like foo.bar into the proper nesting
+// in a nested map/slice structure.
+func (f ConfigBlacklistSettings) Unpack(to interface{}) error {
+	m, ok := to.(map[string]interface{})
+	if !ok {
+		return fmt.Errorf("wrong type, expect map")
+	}
+
+	var expand func(key string, value interface{})
+
+	expand = func(key string, value interface{}) {
+		switch v := value.(type) {
+		case map[string]interface{}:
+			for k, val := range v {
+				expand(fmt.Sprintf("%v.%v", key, k), val)
+			}
+		case []interface{}:
+			for i := range v {
+				expand(fmt.Sprintf("%v.%v", key, i), v[i])
+			}
+		default:
+			m[key] = fmt.Sprintf("%s", value)
+		}
+	}
+
+	for k, val := range m {
+		expand(k, val)
+	}
+	return nil
+}
+
+// NewConfigBlacklist filters configs from CM according to a given blacklist
+func NewConfigBlacklist(patterns ConfigBlacklistSettings) (*ConfigBlacklist, error) {
+	list := ConfigBlacklist{
+		patterns: map[string]*regexp.Regexp{},
+	}
+
+	for field, pattern := range patterns {
+		exp, err := regexp.Compile(pattern)
+		if err != nil {
+			return nil, errors.Wrap(err, fmt.Sprintf("Given expression is not a valid regexp: %s", pattern))
+		}
+
+		list.patterns[field] = exp
+	}
+
+	return &list, nil
+}
+
+// Filter returns a copy of the given ConfigBlocks with the
+func (c *ConfigBlacklist) Filter(configBlocks api.ConfigBlocks) api.ConfigBlocks {
```

Comment:
I'd prefer to keep the ConfigBlock api a as small as possible. Reading the code, I understand that `ConfigBlock` is be mostly a predicate.

e.g. 

```
type Checker interface {
  Check(string, api.ConfigBlock) bool // return false if the config block is rejected by the checker
}

func FilterBlocks(checker Checker, blocks api.ConfigBlocks) api.ConfigBlocks {
  ...
}

func (c *ConfigBlacklist) Check(type string, blk api.ConfigBlock) bool {
  return !c.isBlacklisted(type, blk)
}
```

We could also wrap the Checker to log/collect all config blocks that do not match for reporting back to kibana.


As all functionality is in the management package, do we even need to export the filter function?

Reply:
Yes it's better approach.

I would say no, if we encapsulate all the checks (whitelist+blacklist) and we just check if the config is is valid that removes the need do any filter and this would probably simplify things to know what state the beats is currently in.

All config block valid => config applied 
Any invalid block => do not apply the config.

Reply:
I also thing validate should check all the checks and not fails on the first one so we can report all the errors at once in the UI.

Reply:
+1 on collecting all errors. Use multierror.

Reply:
Not sure I understand the keep `ConfigBlock` API small comment, I don't add anything to ConfigBlock here, as `isBlacklisted` is part of `ConfigBlacklist`.

Do we agree then on treating any a single blacklisted block as a global error that will stop applying new configs? I can do the changes and introduce multierror